### PR TITLE
[SPARK-24542] [SQL] UDF series UDFXPathXXXX allow users to pass carefully crafted XML to access arbitrary files

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?

UDF series UDFXPathXXXX allow users to pass carefully crafted XML to access arbitrary files. Spark does not have built-in access control. When users use the external access control library, users might bypass them and access the file contents.

This PR basically patches the Hive fix to Apache Spark. https://issues.apache.org/jira/browse/HIVE-18879

## How was this patch tested?

A unit test case

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @petermaxlee @rxin @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/UDFXPathUtilSuite.scala [hunk 1]
```diff
@@ -77,6 +77,23 @@ class UDFXPathUtilSuite extends SparkFunSuite {
...
+    val secretValue = String.valueOf(Math.random)
+    val tempFile = File.createTempFile("verifyembed", ".tmp")
+    tempFile.deleteOnExit()
+    val fname = tempFile.getAbsolutePath
+
+    FileUtils.writeStringToFile(tempFile, secretValue)
+
+    val xml =
+      "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n" + "<!DOCTYPE test [ \n" +
+      "    <!ENTITY embed SYSTEM \"" + fname + "\"> \n" + "]>\n" + "<foo>&embed;</foo>"
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/UDFXPathUtilSuite.scala]: can we use multiline string to make it easier to read?

Comment (gatorsmile) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/UDFXPathUtilSuite.scala]: xml has a unique syntax. A little bit hard to make it work sometimes.

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@HyukjinKwon is it possible that the constant build failure is caused by the java style checker? Other PRs that don't touch java files are fine.

**Comment by HyukjinKwon** [comment]
@cloud-fan, I will take a look tonight (singapore timezone). Please feel free to disable it for now to unblock other PRs if you think so. From a very quick look (it's mobile), I think it needs some time.

**Comment by HyukjinKwon** [comment]
you can commwnt this line https://github.com/HyukjinKwon/spark/blob/master/dev/run-tests.py#L577 and add `pass` instead.

**Comment by HyukjinKwon** [comment]
If you think it's not super urgent, please give me few days. I have some speculations about it. Will test it and try a fix tomorrow as soon as we can access to Jenkins results.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

**Pull request closed without merging**